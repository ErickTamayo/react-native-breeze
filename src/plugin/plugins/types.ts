import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import Config from "../utils/config";

type PatterCallableParams = {
  keys: typeof Config.keys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof Config.theme;
  color: typeof Config.color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
