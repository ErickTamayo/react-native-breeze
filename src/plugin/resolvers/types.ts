export type ResolverFunctionParams = {
  input: string;
  groups: { [group: string]: string };
  theme: (key: string) => any;
};

export type ResolverFunction = (params: ResolverFunctionParams) => object;

export type Resolver = {
  pattern: RegExp;
  fn: ResolverFunction;
};
