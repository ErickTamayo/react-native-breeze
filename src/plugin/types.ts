export type Configuration = {
  theme: {
    [key: string]: any;
  };
  variants: {
    [key: string]: any;
  };
  corePlugins: {
    [key: string]: boolean;
  };
  plugins: string[];
};
