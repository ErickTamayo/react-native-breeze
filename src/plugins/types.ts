import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import { keys, colorKeys, theme, color } from "../config";

type PatterCallableParams = {
  keys: typeof keys;
  colorKeys: typeof colorKeys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof theme;
  color: typeof color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
