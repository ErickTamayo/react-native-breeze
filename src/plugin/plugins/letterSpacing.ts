import { validate } from "../utils/validation";
import { PluginFunction, PluginPattern } from "./types";

export type PluginGroups = {
  key: string;
};

export const pattern: PluginPattern = ({ keys }) => {
  const spacingKeys = keys("letterSpacing", "|");
  return new RegExp(`^tracking-(?<key>${spacingKeys})$`);
};

export const plugin: PluginFunction<PluginGroups> = ({
  input,
  groups,
  theme,
}) => {
  const { key } = groups;
  const value = theme<number>(["letterSpacing", key]);

  if (!validate(input, value, ["number"])) return {};

  return { letterSpacing: value };
};
