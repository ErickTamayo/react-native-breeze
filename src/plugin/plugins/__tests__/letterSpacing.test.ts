import { pattern, plugin, PluginGroups } from "../letterSpacing";
import { PatternCallable, PluginFunctionReturnType } from "../types";

describe("letterSpacing", () => {
  it("Should match the correct pattern", () => {
    const theme = jest.fn().mockReturnValue({ normal: 0, wider: 0.8 });
    const callable = pattern as PatternCallable;

    expect(callable({ theme }).exec("tracking-normal")).toBeTruthy();
    expect(callable({ theme }).exec("tracking-wider")).toBeTruthy();

    expect(callable({ theme }).exec("tracking-unknown")).toBeFalsy();
  });

  // prettier-ignore
  it.each<[input: string, letterSpacing: { [key: string]: number }, returned: number, expected: PluginFunctionReturnType]>([
    ["tracking-normal", { normal: 0, wider: 0.8 }, 0, { letterSpacing: 0 }],
    ["tracking-wider", { normal: 0, wider: 0.8 }, 0.8, { letterSpacing: 0.8 }],
   ])(
    "Should parse (%s) correctly",
    (input, radii, returned, expected) => {
      const theme = jest.fn()

      theme.mockReturnValueOnce(radii);
      const callable = pattern as PatternCallable;
      const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;

      theme.mockReturnValueOnce(returned);
      const result = plugin({ input, groups, theme });

      expect(result).toEqual(expected);
    }
  );

  it("Should log error if the letterSpacing value is not valid", () => {
    const theme = jest.fn();

    theme.mockReturnValueOnce({ normal: "0" });
    const input = "tracking-normal";
    const callable = pattern as PatternCallable;
    const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;
    const spy = jest.spyOn(console, "error").mockImplementation(() => {});

    theme.mockReturnValueOnce("0");
    const result = plugin({ input, groups, theme });

    expect(result).toEqual({});
    expect(spy).toHaveBeenCalled();

    spy.mockRestore();
  });
});
