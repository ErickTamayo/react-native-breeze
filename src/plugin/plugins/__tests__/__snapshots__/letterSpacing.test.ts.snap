// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`letterSpacing Should parse (tracking-normal) correctly 1`] = `
Object {
  "letterSpacing": 0,
}
`;

exports[`letterSpacing Should parse (tracking-tight) correctly 1`] = `
Object {
  "letterSpacing": -0.4,
}
`;

exports[`letterSpacing Should parse (tracking-tighter) correctly 1`] = `
Object {
  "letterSpacing": -0.8,
}
`;

exports[`letterSpacing Should parse (tracking-wide) correctly 1`] = `
Object {
  "letterSpacing": 0.4,
}
`;

exports[`letterSpacing Should parse (tracking-wider) correctly 1`] = `
Object {
  "letterSpacing": 0.8,
}
`;

exports[`letterSpacing Should parse (tracking-widest) correctly 1`] = `
Object {
  "letterSpacing": 1.6,
}
`;
